aten/src/ATen/native/cuda/ForeachBinaryOpScalar.cu

```diff
@@ -58,16 +58,53 @@ void foreach_tensor_##NAME##_scalar_kernel_cuda_(TensorList tensors, Scalar scal
                                                                                                     \
 std::vector<Tensor> foreach_tensor_##NAME##_scalar_kernel_cuda(TensorList tensors, Scalar scalar) { \
     check_foreach_api_restrictions(tensors);                                                        \
-    if (!can_use_fast_route({tensors}, {scalar})) {                                                 \
+    if (!can_use_fast_route(tensors, scalar, DIVISION_OP)) {                                        \
         return at::native::foreach_tensor_##NAME##_scalar_kernel_slow(tensors, scalar);             \
     }                                                                                               \
                                                                                                     \
     return foreach_binary_op<OP>(tensors, scalar);                                                  \
 }
 
-FOREACH_BINARY_OP_SCALAR(add, std::plus);
-FOREACH_BINARY_OP_SCALAR(sub, std::minus);
-FOREACH_BINARY_OP_SCALAR(mul, std::multiplies);
-FOREACH_BINARY_OP_SCALAR(div, std::divides);
+FOREACH_BINARY_OP_SCALAR(add, std::plus, false);
+FOREACH_BINARY_OP_SCALAR(mul, std::multiplies, false);
+
+// In the case of division, integer inputs will result in float. 
+// Currently multi tensor apply can only return result of the same type as input.
+FOREACH_BINARY_OP_SCALAR(div, std::divides, true);
+
+// In the case of subtraction, we dont allow scalar to be boolean following the torch.sub logic
+void foreach_tensor_sub_scalar_kernel_cuda_(TensorList tensors, Scalar scalar) {
+    check_foreach_api_restrictions(tensors);
+
+    TORCH_CHECK(tensors[0].scalar_type() != kBool || !scalar.isBoolean(),
+              "Subtraction, the `-` operator, with two bool tensors is not supported."
+              "Use the `^` or `logical_xor()` operator instead.")
+    TORCH_CHECK(tensors[0].scalar_type() != kBool && !scalar.isBoolean(),
+              "Subtraction, the `-` operator, with a bool tensor is not supported. "
+              "If you are trying to invert a mask, use the `~` or `logical_not()` operator instead.");
```

Comment:
As I mentioned in the last PR in the stack, we should really refactor these messages out so that they're not being repeated so much